src/libcharon/plugins/kernel_super/kernel_super_ipsec.c

```diff
@@ -174,7 +174,116 @@ static status_t del_routes_from_routing_table(private_kernel_super_ipsec_t* this
+	oper_data.activeness_status = ike_sa->get_is_active(ike_sa);
+
+	char device_public_addr[MAX_IP_STR_LEN + 1];
+	char remote_public_addr[MAX_IP_STR_LEN + 1];
+	char local_network[MAX_IP_PREFIX_STR_LEN +1];
+	data->src->address_to_string(data->src, &device_public_addr[0]);
+	data->src->address_to_string(data->dst, &remote_public_addr[0]);
+	ts_to_prefix_string_for_logs(id->dst_ts, &local_network[0]);
+
+	oper_data.device_public_address = &device_public_addr[0];
+	oper_data.remote_public_address = &remote_public_addr[0];
+	oper_data.device_public_src_port = data->src->get_port(data->src);
+	oper_data.local_network = &local_network[0];
+
+	oper_data.account_id = super_peer_info->get_account_id(super_peer_info);
+	oper_data.vrf_name = super_peer_info->get_vrf_name(super_peer_info);
+	oper_data.device_id = super_peer_info->get_device_id(super_peer_info);
+
+	if(! oper_data.account_id || ! oper_data.vrf_name || !oper_data.device_id) {
+		LOG(DBG_KNL, SL_LOG_ERR, "super_peer_info has NULL values account-id: %s, vrf-name: %s, device-id: %s", oper_data.account_id, oper_data.vrf_name, oper_data.device_id);
+		goto out;
+	}
+
+	struct pm_terminator_conn* terminator_conn = NULL;
+	terminator_conn = lookup_terminator_conn(this, _get_sp_lookup_addr(data, id));
+	oper_data.term_id = picomsg_terminator_conn_get_id(terminator_conn);
+	picomsg_terminator_conn_unref(terminator_conn);
+
+	const struct system_enterprise_map_lookup_params params = {
+		.terminator_id = oper_data.term_id,
+		.tenant_id = super_peer_info->get_tenant_id(super_peer_info),
+		.account_id = oper_data.account_id,
+		.vrf_name = oper_data.vrf_name,
+	};
+
+	struct system_enterprise_map_enterprise enterprise;
+	if (! system_enterprise_map_lookup_enterprise(this->system_enterprise_map, &params, &enterprise)) {
+		LOG(DBG_KNL, SL_LOG_ERR, "peer information (terminator_id=%u, tenant_id=%s, account_id=%s, vrf_name=%s) doesn't exist in system info map",
+			oper_data.term_id, super_peer_info->get_tenant_id(super_peer_info), oper_data.account_id, oper_data.vrf_name);
+			goto out;
+	}
+	oper_data.neigh_sub_interface = enterprise.neigh_sub_interface;
+	if (! oper_db_writer_set_tunnel_oper_data(oper_data, write_uptime)) {
+		goto out;
+	}
+	return;
+	out:
+	LOG(DBG_KNL, SL_LOG_ERR, "Failed to write tunnel");
+}
 
+static void delete_terminator_tunnels_oper(ike_super_peer_info_t* super_peer_info, uint32_t terminator_id) {
+
+	if(IKE_SUPER_PEER_INFO_TYPE_TUNNEL == super_peer_info->get_type(super_peer_info)) {
+		return;
+	}
+	if(! oper_db_writer_remove_tunnel_oper_data(super_peer_info->get_vrf_name(super_peer_info), super_peer_info->get_device_id(super_peer_info))) {
+		LOG(DBG_KNL, SL_LOG_ERR, "Failed to remove tunnel_oper_data vrf_name: %s device_id: %s",
+			super_peer_info->get_vrf_name(super_peer_info), super_peer_info->get_device_id(super_peer_info));
+	}
+
+	if (! oper_db_writer_del_terminator_vrf_tunnel_counter(terminator_id, super_peer_info->get_tenant_id(super_peer_info),
```

Comment:
There can be multiple tunnels using the same terminator/tenant/account/vrf
You shouldn't delete the whole counter field in case a single tunnel was down. This should happen only when the tenant/account/vrf is **explicit removed** (keep in mind that `write_terminator_tunnels_oper()` can concurrently run and write to the same tenant/account/vrf_name counter field just after you deleted the enterprise/VRF, so you need to synchronize with that timer)